### ISSUE TITLE
Multiple disks specified in clone don't work and VM doesn't power up

### ISSUE THREAD

**Issue by Tigershark2005** [hashicorp/terraform-provider-vsphere#1578]
### Terraform Version

v1.1.4

### vSphere Provider Version

2.0.2

### vSphere Version

7.0.3, also tried on 6.7

### Affected Resource(s)

`vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
resource "vsphere_virtual_machine" "cloned_virtual_machine" {
    name = var.vsphere_virtual_machine_name
    resource_pool_id = data.vsphere_resource_pool.pool.id
    datastore_id = data.vsphere_datastore.datastore.id
    depends_on = [
      data.vsphere_resource_pool.pool
    ]

    num_cpus = var.vsphere_num_cpus
    cpu_hot_add_enabled = true
    memory = var.vsphere_memory
    memory_hot_add_enabled = true
    guest_id = data.vsphere_virtual_machine.template.guest_id
    folder = var.vsphere_folder
    scsi_type = data.vsphere_virtual_machine.template.scsi_type

    network_interface {
      network_id = data.vsphere_network.network.id
      adapter_type = "vmxnet3"
    }

    
    dynamic "disk" {
        for_each = var.vsphere_disks
        content {
            label = "disk${disk.value["number"]}"
            size = disk.value["size"]
            unit_number = disk.value["number"]
        }
    }

    clone {
        template_uuid = data.vsphere_virtual_machine.template.id

        customize {
          network_interface {
            ipv4_address = var.vsphere_ipaddress
            ipv4_netmask = var.vsphere_netmask
            dns_server_list = var.vsphere_dns_servers
          }
          ipv4_gateway = var.vsphere_ipgateway

          windows_options {
            computer_name = var.vsphere_virtual_machine_name
            admin_password = "-------"
            join_domain = "removed"
            domain_admin_user = var.domainadminuser
            domain_admin_password = var.domainadminpassword
            time_zone = 35
          }
          timeout = 30  //some VMs hang on "Getting Started" in Windows and need this extended timeout

        }
    }    
}
```
I'm using a dynamic block but the tf plan shows the expected attributes that will be used. Disk size, provisioning style, all parameters are correct, so it's nothing to do with the dynamic block. I could have just typed the below array into the module. This is the block I'm sending in for disk:
```
vsphere_disks = [
        {
            number = 0
            size = 75
        },
        {
            number = 1
            size = 100
        }
]
```
### Debug Output

No negative output except "Unable to enumerate all disks." That's because the second hard drive has size 0MB and a different provisioning type and the VM can't startup.

### Panic Output

No panic

### Expected Behavior

The VM should have had both hard drives as defined in the code. Disk 1, 75GB, Disk 2, 100GB

### Actual Behavior

First disk is created as expected, second disk is 0MB and thick provisioned lazy zeroed

### Steps to Reproduce

`terraform apply` which will show the correct values to be used for creation

### Important Factoids

I'm using local execution.

### References

* #0000

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

**Comment by tenthirtyam** [comment]
Taking a _very_ quick look at the issue, I see that `firmware` and `enable_secure_boot` are not defined, both of which could be pulled from the template with a data source (`data.vsphere_virtual_machine.template.firmware`). If your template uses EFI / EFI Secure Boot, but these are not defined / sourced it will fail to boot as it defaults to BIOS.

Ryan

**Comment by Tigershark2005** [comment]
> Taking a _very_ quick look at the issue, I see that `firmware` and `enable_secure_boot` are not defined, both of which could be pulled from the template with a data source (`data.vsphere_virtual_machine.template.firmware`). If your template uses EFI / EFI Secure Boot, but these are not defined / sourced it will fail to boot as it defaults to BIOS.
> 
> Ryan

No, the VM creates fine with one disk. We've never set those parameters and they're carried over anyway with the clone.

**Comment by Tigershark2005** [comment]
<script src="https://gist.github.com/Tigershark2005/440f44c987ac043368beaeccdf3683e5.js"></script>

**Comment by tenthirtyam** [comment]
Apologies for the delay. A cursory review of the Gist looks good and nothing major stands out. 

_However,_ can you review your template and ensure that it's not using either EFI and optionally, secure boot in the `VM Options > Boot Options` settings.?  This is the recommended setting (and default) for Windows 2016 and later in vSphere 7.0 (at least). 

If this is enabled, then the issue is most likely that you'll need to add the following

```hcl
resource "vsphere_virtual_machine" "cloned_virtual_machine" {
   #.... other configurations...
   firmware = var.firmware
   enable_secure_boot = var.enable_secure_boot
}
```

or 

```hcl
resource "vsphere_virtual_machine" "cloned_virtual_machine" {
   #.... other configurations...
   firmware = data.vsphere_virtual_machine.template.firmware
   enable_secure_boot = data.vsphere_virtual_machine.template.enable_secure_boot
}
```

The reason for this would be that if EFI is in use, but not set explicitly, then it will default to BIOS and the guest operating system will not boot.

For example, I use EFI in [vmware-samples/packer-examples-for-vsphere ](https://github.com/vmware-samples/packer-examples-for-vsphere) as the default and I **must** add the settings to Terraform or the machine images will default to BIOS and not boot.

https://github.com/hashicorp/terraform-provider-vsphere/blob/ee62ff06fdcc8b0ca6185a03f02fbc0b1430d31f/vsphere/virtual_machine_config_structure.go#L277-L282

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by tenthirtyam** [comment]
Also, to confirm, does the resource not "power on" as mentioned in the issue title, or is it that the guest operating system (WIndows) does not boot after the power on.

**Comment by Tigershark2005** [comment]
But I'm not necessarily worried about boot directly at this point. The core issue is I get one drive that's correct and the second drive is 0MB. So multiple disk definitions aren't working. Not booting I think is a side effect.


________________________________

**Comment by tenthirtyam** [comment]
Can you confirm the settings on the template before progressing further into analysis to rule this concern out?

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by tenthirtyam** [comment]
The issue will be marked as `closed` soon. If there is an update to [the previous comment](https://github.com/hashicorp/terraform-provider-vsphere/issues/1578#issuecomment-1043406241), please let us know.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by Tigershark2005** [comment]
What sort of settings are you needing? This is an existing VM that's being cloned.


________________________________

**Comment by tenthirtyam** [comment]
Can you confirm the virtual firmware settings are BIOS vs EFI? Can you paste a redacted and dropped screenshot?

**Comment by Tigershark2005** [comment]
It's set to BIOS. I've included the screenshot. But I'm still not trying to troubleshoot a boot issue, I'm trying to troubleshoot the errored second hard drive, which wouldn't have anything to do with BIOS settings. 
![image](https://user-images.githubusercontent.com/20572929/157200198-f4b54337-2f3c-42b6-809c-b8fca6ab9251.png)

**Comment by Tigershark2005** [comment]
@tenthirtyam I saw you removed your assignment and the issue is still pending-closure but nothing has been fixed.

**Comment by tenthirtyam** [comment]
re: "The core issue is I get one drive that's correct and the second drive is 0MB"

This would be expected if you are using NFS for additional disks added to the clone as it's thin provisioned by default.

Are you using NFS?

Ryan

**Comment by Tigershark2005** [comment]
I don't believe we're using NFS but I haven't seen anything happen that way in the past. If I set drive 2 to 100GB then that's how big it would be, if it's thin provisioned then it just doesn't use up 100GB on disk.

**Comment by tenthirtyam** [comment]
If time permits this week I can attempt a reproduction of the issue, but the environmentals may differ. 

Ryan

**Comment by Tigershark2005** [comment]
The main thing to check would be passing in two disks, units 0 and 1. My disk 1 shows as 0GB even though the plan says it will be the expected size. Execution doesn't create with the right size.


________________________________

**Comment by tenthirtyam** [comment]
Can you confirm that diakO is the correct size, `>=` the originating template and whether or not the size was increased?

I'll proceed with the assumption that the datastores are backed by VMFS vs MFS.

Ryan

**Comment by Tigershark2005** [comment]
Yes, I'm not sure if you saw my earlier comment but these are vmfs6 backed by FibreChannel disks. Disk0 is correct and creating as expected as well. If memory serves correctly, this worked in version 1.10 or something around there. I can't say with confidence that there weren't any slight adjustments in our code since then but I faintly remember this being a working then not sort of thing so that lead me to believe it was something with the tf engine.

**Comment by tenthirtyam** [comment]
It may be defaulting to thin vs eager zero, would need to check the docs and code to be sure as well as any existing `area/storage` issues.

Ryan

**Comment by Tigershark2005** [comment]
No issues on the storage side, these are solid state arrays running sub-ms latency. I'll check on the type, disk1 is a new disk so there shouldn't be any competing with an existing disk1 type. And we've also done this process of adding a second disk to the clone in this manner before. I'll double check docs too and see if anything stands out to me on types or anything.


________________________________

**Comment by Tigershark2005** [comment]
I haven't found anything related in my search of those tags so far. The template is thin provisioned for disk0 and I'm not specifying that parameter so it's defaulting to thin, and that disk shows up as expected. Below is what I'm sending to the module for disks:
```
vsphere_disks = [
        {
            number = 0
            size = 75
        },
        {
            number = 1
            size = 75
        }
    ]
```
Then in the module it looks like the following:
```
dynamic "disk" {
        for_each = var.vsphere_disks
        content {
            label = "disk${disk.value["number"]}"
            size = disk.value["size"]
            unit_number = disk.value["number"]
        }
    }
```
And unless I'm missing something, the plan shows what I would expect:
```
+ disk {
          + attach            = false
          + controller_type   = "scsi"
          + datastore_id      = "<computed>"
          + device_address    = (known after apply)
          + disk_mode         = "persistent"
          + disk_sharing      = "sharingNone"
          + eagerly_scrub     = false
          + io_limit          = -1
          + io_reservation    = 0
          + io_share_count    = 0
          + io_share_level    = "normal"
          + keep_on_remove    = false
          + key               = 0
          + label             = "disk0"
          + path              = (known after apply)
          + size              = 75
          + storage_policy_id = (known after apply)
          + thin_provisioned  = true
          + unit_number       = 0
          + uuid              = (known after apply)
          + write_through     = false
        }
      + disk {
          + attach            = false
          + controller_type   = "scsi"
          + datastore_id      = "<computed>"
          + device_address    = (known after apply)
          + disk_mode         = "persistent"
          + disk_sharing      = "sharingNone"
          + eagerly_scrub     = false
          + io_limit          = -1
          + io_reservation    = 0
          + io_share_count    = 0
          + io_share_level    = "normal"
          + keep_on_remove    = false
          + key               = 0
          + label             = "disk1"
          + path              = (known after apply)
          + size              = 75
          + storage_policy_id = (known after apply)
          + thin_provisioned  = true
          + unit_number       = 1
          + uuid              = (known after apply)
          + write_through     = false
        }
```
Near the end of creation I can see the first disk is fine and the second one is at 0GB. After it fails the VM is of course deleted so I can't check anything further there.

**Comment by Tigershark2005** [comment]
Ok I enabled debugging in the vsphere provider and I'm seeing the following. Keep in mind there only seems to be an issue when I'm specifying two disks and doing it with one at the clone's value of 75GB works fine. I see the xml send to add unit 1 but nothing for 0, so it wouldn't be changing unit 0/disk0. However, it's returning this error:
```

```
<key>com.vmware.vim.vpxd.vpx.vmprov.ReconfigureVm</key>
<message>Reconfiguring Virtual Machine on destination host</message>
</description>
<name>ReconfigVM_Task</name>
<descriptionId>VirtualMachine.reconfigure</descriptionId>
<entity type="VirtualMachine">vm-3385</entity>
<entityName>xxxxxxxxx3</entityName>
<state>error</state>
<cancelled>false</cancelled>
<cancelable>false</cancelable>
-<error>
-<fault xsi:type="InvalidDeviceSpec">
<property>VirtualDisk.capacityInKB</property>
<deviceIndex>0</deviceIndex>
</fault>
<localizedMessage>Invalid configuration for device '0'.</localizedMessage>
</error>
```
So it's wanting to create disk1 (unitNumber 1), and may actually be getting some amount through that piece though I can't tell since this edit happens in milliseconds. But the xml return refers to the disk0's size. Being that this works fine when only specifying one disk this seems like a provider issue to me.

**Comment by 8uachaille** [comment]
+1 
I believe I am also seeing this problem per https://github.com/hashicorp/terraform-provider-vsphere/issues/1639

**Comment by 8uachaille** [comment]
@Tigershark2005  -  

I had many of the same symptoms as you and just got a build to work with multiple filesystems as per  https://github.com/hashicorp/terraform-provider-vsphere/issues/1639

I am no expert but I might be willing to compare configs if you are interested and we can do so quickly

**Comment by cloudypc** [comment]
+1 
Also seeing this issue

**Environment**
vCenter 7.0.3 build 20395099
vSphere Client version 7.0.3.00700
Terraform vSphere provider vsphere v2.2.0
Terraform version v1.3.5

Is this a valid bug that is under review? Unsure as the issue looks to be open but there has not been a lot of activity lately.

**Comment by tenthirtyam** [comment]
HI @cloudypc, this state of this issue can be seen in the assigned milestone which is currently set to the `backlog`.

![image](https://user-images.githubusercontent.com/7771363/203211353-78b87ff4-d3be-4f53-a107-b0eb59c4fb41.png)

Also, please consider the **Community Note** in the issue description. The issue only has one 👍 reaction.

> ### Community Note
> * Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
> * Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
> * If you are interested in working on this issue or have submitted a pull request, please leave a comment.

To be completely transparent, I was on sabbatical recently and have lately only bee performing initial issue triage and pull request reviews for some upcoming features and haven't had much spare cycles to dive deeper, which is often in my spare personal time.

Please give the issue a 👍 and if you'd like to triage and contribute a fix, we'd be happy to review.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by Flawke** [comment]
It happens weather cloning a VM or creating a new one with multiple disks.
If you create\clone with a single disk it works fine. However, any additional disk will cause this issue.
The vmdk is not being created at all yet the vmx file point to that non-existing vmdk and throwing the error.

For some reason, the provider does not create any additional vmdks.

**Comment by Zheer09** [comment]
We have the same issue but only from pipeline , i when i run it from my machine all fine , but in the pipeline same error will show a side note is that we install the vsphere provider manually to the server version: 2.4.0 this is the `.terraform.d/plugins/registry.terraform.io/hashicorp/vsphere/2.4.0/linux_386/terraform-provider-vsphere_v2.4.0_x5` I don't know how this is related but my machine and runner pipeline version are exactly the same the only different is that i use linux_am64 , runner use linux_386.

**Comment by mdinizcloud** [comment]
I have the same issues using powershell, using shell from centos7 it's work!

**Comment by frankyie2022** [comment]
Any updates? Having the same issue. Using a mware template with one disk, trying to extra vmdk disks.

Thx

**Comment by marayas** [comment]
This also happens when creating the disks and trying to attach them to the VM, if the same number of disks are already defined in the template works fine, adding an extra disk gives the issue.

**Comment by JRodes34** [comment]
Hello, I had been dealing with this issue for a couple weeks. I was browsing all the forums I could, but no one was able to give an answer.

Well yesterday, I figured it out!

It ended up being the fact that I was using a 32 bit version of Terraform. I upgraded to the 64 bit version, and the issue went away.

Not sure if anyone is still looking at this thread, but I hope this helps.
